### PR TITLE
Add diff op

### PR DESCRIPTION
**Pull request by andyjiang1116**
### PR types

New features
### PR changes

OPs
### Describe

add `paddle.diff()` OP，support CPU and GPU.    

截屏2021-11-30 上午9 30 34
截屏2021-11-30 上午9 30 44

### CODE REVIEW

**Review by MissPenguin — DISMISSED** [review]

**Review by fuyinno4 — DISMISSED** [review]

File: python/paddle/tensor/math.py [hunk 1]
```diff
@@ -2611,3 +2611,166 @@ def atan2(x, y, name=None):
         helper.append_op(
                 type='atan2', inputs=inputs, outputs={'Out': out})
         return out
+
+
+def diff(x, n=1, dim=-1, prepend=None, append=None, name=None):
```
Comment (jeff41404) [inline python/paddle/tensor/math.py]: in order to align to other apis, we shall use axis instand of dim

**Comment by jeff41404** [comment]
the 3rd example's output should be [[3, 3, 3]] ?

**Comment by andyjiang1116** [comment]
> the 3rd example's output should be [[3, 3, 3]] ?

ok, ths, I'll fix it

**Review by jeff41404 — DISMISSED** [review]
lgtm

**Review by TCChenlong — APPROVED** [review]
LGTM

File: python/paddle/tensor/math.py [hunk 2]
```diff
@@ -2611,3 +2611,166 @@ def atan2(x, y, name=None):
...
+    r"""
+    Computes the n-th forward difference along the given axis.
+    The first-order differences is computed by using the following formula: 
+
+    .. math::
+
+        out[i] = x[i+1] - x[i]
+    
+    Higher-order differences are computed by using paddle.diff() recursively. 
+    Only n=1 is currently supported.
```
Comment (TCChenlong) [inline python/paddle/tensor/math.py:2626]: 这里用 note 的格式，加强提醒。

Comment (andyjiang1116) [reply python/paddle/tensor/math.py:2626]: 好的，谢谢~

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Pull request merged**